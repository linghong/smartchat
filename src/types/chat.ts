export interface Chat {
  id: number;
  title: string;
}

export type ChatRole = 'system' | 'assistant' | 'user';

export type ChatType = {
  role: ChatRole;
  content: string;
};

export interface ImageFile {
  base64Image: string;
  mimeType: string;
  size: number;
  name: string;
  width?: number;
  height?: number;
}

export interface FileData {
  base64Content: string;
  type: string;
  size: number;
  name: string;
}

export interface Message {
  question: string;
  answer: string;
  model: string;
}

export interface OpenAIChatContentImage {
  type: 'image_url';
  image_url: {
    url: string;
    detail: 'auto' | 'low' | 'high';
  };
}
export interface OpenAIChatContentText {
  type: 'text';
  text: string;
}
