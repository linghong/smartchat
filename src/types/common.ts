export interface OptionType {
  value: string;
  label: string;
  tags?: string[];
  category?: string;
  contextWindow?: number;
  vision?: boolean;
}

export interface ChatOption extends OptionType {
  tags: string[];
}

export type ApiResponse = {
  message: string;
  fileName: string;
  error?: string;
};

export type InputErrors = {
  [key: string]: string | null;
};

export type Input<T> = {
  [key: string]: T;
};

export type InputData = {
  [key: string]: string | number | null;
};

export type UploadData = {
  [key: string]: File | null;
};

export type UploadErrors = {
  [key: string]: string | null;
};
