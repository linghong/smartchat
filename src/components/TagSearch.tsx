import React, { useState, useEffect, useMemo } from 'react';
import { Search, X } from 'lucide-react';
import { OptionType } from '@/src/types/common';

interface TagSearchProps {
  chats: OptionType[];
  onFilterChats: (filteredChats: OptionType[]) => void;
}

const TagSearch: React.FC<TagSearchProps> = ({ chats, onFilterChats }) => {
  const [searchTerm, setSearchTerm] = useState('');
  const [selectedTags, setSelectedTags] = useState<string[]>([]);
  const [suggestedTags, setSuggestedTags] = useState<string[]>([]);

  const allTags = useMemo(() => {
    return Array.from(new Set(chats.flatMap(chat => chat.tags || [])));
  }, [chats]);

  useEffect(() => {
    const filteredChats = chats.filter(chat =>
      selectedTags.every(tag => chat.tags?.includes(tag))
    );
    onFilterChats(filteredChats);

    const newSuggestedTags = allTags.filter(
      tag =>
        tag.toLowerCase().includes(searchTerm.toLowerCase()) &&
        !selectedTags.includes(tag)
    );
    setSuggestedTags(newSuggestedTags);
  }, [allTags, chats, onFilterChats, searchTerm, selectedTags]);

  const handleTagClick = (tag: string) => {
    setSelectedTags([...selectedTags, tag]);
    setSearchTerm('');
  };

  const handleTagRemove = (tag: string) => {
    setSelectedTags(selectedTags.filter(t => t !== tag));
  };

  return (
    <div className="mb-6 bg-white p-4 rounded-lg shadow-md">
      <label
        htmlFor="tag-search"
        className="block text-sm font-medium text-gray-700 mb-2"
      >
        Search Tags
      </label>
      <div className="relative">
        <input
          id="tag-search"
          type="text"
          value={searchTerm}
          onChange={e => setSearchTerm(e.target.value)}
          placeholder="Type to search tags, and click a tag to select"
          className="w-full px-4 py-2 border border-gray-300 rounded-md pl-10 text-sm focus:ring-2 focus:ring-blue-500 focus:border-blue-500"
          aria-label="Search for tags"
        />
        <Search
          className="absolute left-3 top-1/2 transform -translate-y-1/2 text-gray-400"
          size={16}
          aria-hidden="true"
        />
      </div>
      {selectedTags.length > 0 && (
        <div className="mt-3">
          <h3 className="text-sm font-medium text-gray-700 mb-2">
            Selected Tags:
          </h3>
          <div className="flex flex-wrap gap-2">
            {selectedTags.map(tag => (
              <span
                key={tag}
                className="bg-blue-100 text-blue-800 px-3 py-1 rounded-full text-sm flex items-center"
              >
                {tag}
                <button
                  onClick={() => handleTagRemove(tag)}
                  className="ml-2 text-blue-600 hover:text-blue-800 focus:outline-none focus:ring-2 focus:ring-blue-500 rounded-full"
                  aria-label={`Remove ${tag} tag`}
                >
                  <X size={14} />
                </button>
              </span>
            ))}
          </div>
        </div>
      )}
      {suggestedTags.length > 0 && (
        <div className="mt-4">
          <h3 className="text-sm font-medium text-gray-700 mb-2">
            Suggested Tags:
          </h3>
          <div className="flex flex-wrap gap-2">
            {suggestedTags.map(tag => (
              <button
                key={tag}
                onClick={() => handleTagClick(tag)}
                className="bg-gray-100 hover:bg-gray-200 text-gray-800 px-3 py-1 rounded-full text-sm transition-colors duration-200 focus:outline-none focus:ring-2 focus:ring-blue-500"
              >
                {tag}
              </button>
            ))}
          </div>
        </div>
      )}
    </div>
  );
};

export default TagSearch;
