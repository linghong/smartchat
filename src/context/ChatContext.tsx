'use client';
import React, {
  createContext,
  useContext,
  useState,
  ReactNode,
  Dispatch,
  SetStateAction
} from 'react';

import { Message, FileData } from '@/src/types/chat';
import { OptionType } from '@/src/types/common';
import { initialMessage, defaultModel } from '@/src/utils/initialData';

export interface ChatContextType {
  isNewChat: boolean;
  setIsNewChat: Dispatch<SetStateAction<boolean>>;
  isConfigPanelVisible: boolean;
  setIsConfigPanelVisible: Dispatch<SetStateAction<boolean>>;
  isSearchChat: boolean;
  setIsSearchChat: Dispatch<SetStateAction<boolean>>;
  fileSrcHistory: FileData[][];
  setFileSrcHistory: Dispatch<SetStateAction<FileData[][]>>;
  chatHistory: Message[];
  setChatHistory: Dispatch<SetStateAction<Message[]>>;
  activeChat: OptionType;
  setActiveChat: Dispatch<SetStateAction<OptionType>>;
  chats: OptionType[];
  setChats: Dispatch<SetStateAction<OptionType[]>>;
  selectedModel: OptionType;
  setSelectedModel: Dispatch<SetStateAction<OptionType>>;
}

export const ChatContext = createContext<ChatContextType | undefined>(
  undefined
);

export const ChatProvider: React.FC<{ children: ReactNode }> = ({
  children
}) => {
  const [isNewChat, setIsNewChat] = useState(true);
  const [isConfigPanelVisible, setIsConfigPanelVisible] = useState(false);
  const [isSearchChat, setIsSearchChat] = useState(false);
  const [fileSrcHistory, setFileSrcHistory] = useState<FileData[][]>([[]]);
  const [chatHistory, setChatHistory] = useState<Message[]>([initialMessage]);
  const [activeChat, setActiveChat] = useState<OptionType>({
    value: '0',
    label: '0',
    tags: []
  }); //new chat
  const [chats, setChats] = useState<OptionType[]>([]);
  const [selectedModel, setSelectedModel] = useState<OptionType>(defaultModel);

  return (
    <ChatContext.Provider
      value={{
        isNewChat,
        setIsNewChat,
        isConfigPanelVisible,
        setIsConfigPanelVisible,
        isSearchChat,
        setIsSearchChat,
        fileSrcHistory,
        setFileSrcHistory,
        chatHistory,
        setChatHistory,
        activeChat,
        setActiveChat,
        chats,
        setChats,
        selectedModel,
        setSelectedModel
      }}
    >
      {children}
    </ChatContext.Provider>
  );
};

export const useChatContext = () => {
  const context = useContext(ChatContext);
  if (context === undefined) {
    throw new Error('useChatContext must be used within a ChatProvider');
  }
  return context;
};
