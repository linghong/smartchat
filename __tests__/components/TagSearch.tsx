import React from 'react';
import { render, screen, fireEvent, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';

import TagSearch from '@/src/components/TagSearch';
import { renderWithContext } from '@/__tests__/test_utils/context';
import { OptionType } from '@/src/types/common';

const mockOnFilterChats = jest.fn();

const defaultProps = {
  onFilterChats: mockOnFilterChats
};

describe('TagSearch Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should render the search input', () => {
    renderWithContext(<TagSearch {...defaultProps} />);
    expect(screen.getByLabelText('Search for tags')).toBeInTheDocument();
  });

  it('should update search term when typing', () => {
    renderWithContext(<TagSearch {...defaultProps} />);
    const input = screen.getByLabelText('Search for tags');
    fireEvent.change(input, { target: { value: 'test' } });
    expect(input).toHaveValue('test');
  });

  it('should display suggested tags based on search term', async () => {
    const mockChats: OptionType[] = [
      { value: '1', label: 'Chat 1', tags: ['tag1', 'tag2'] },
      { value: '2', label: 'Chat 2', tags: ['tag2', 'tag3'] }
    ];

    renderWithContext(<TagSearch {...defaultProps} />, { chats: mockChats });

    const input = screen.getByLabelText('Search for tags');
    fireEvent.change(input, { target: { value: 'tag' } });

    await waitFor(() => {
      expect(screen.getByText('tag1')).toBeInTheDocument();
      expect(screen.getByText('tag2')).toBeInTheDocument();
      expect(screen.getByText('tag3')).toBeInTheDocument();
    });
  });

  it('should add a tag when clicked', async () => {
    const mockChats: OptionType[] = [
      { value: '1', label: 'Chat 1', tags: ['tag1', 'tag2'] }
    ];

    renderWithContext(<TagSearch {...defaultProps} />, { chats: mockChats });

    const input = screen.getByLabelText('Search for tags');
    fireEvent.change(input, { target: { value: 'tag' } });

    await waitFor(() => {
      fireEvent.click(screen.getByText('tag1'));
    });

    expect(screen.getByText('Selected Tags:')).toBeInTheDocument();
    expect(screen.getByText('tag1')).toBeInTheDocument();
  });

  it('should remove a tag when remove button is clicked', async () => {
    const mockChats: OptionType[] = [
      { value: '1', label: 'Chat 1', tags: ['tag1', 'tag2'] }
    ];

    renderWithContext(<TagSearch {...defaultProps} />, { chats: mockChats });

    const input = screen.getByLabelText('Search for tags');
    fireEvent.change(input, { target: { value: 'tag' } });

    await waitFor(() => {
      expect(screen.getByText('tag1')).toBeInTheDocument();
    });

    fireEvent.click(screen.getByText('tag1'));

    // Verify the tag is added to selected tags
    await waitFor(() => {
      expect(screen.getByText('Selected Tags:')).toBeInTheDocument();
      expect(screen.getByText('tag1')).toBeInTheDocument();
    });

    const removeButton = screen.getByLabelText('Remove tag1 tag');
    fireEvent.click(removeButton);

    // Wait for the tag to be removed
    await waitFor(() => {
      expect(screen.queryByText('Selected Tags:')).not.toBeInTheDocument();
      expect(
        screen.queryByRole('button', { name: 'Remove tag1 tag' })
      ).not.toBeInTheDocument();
    });

    // Verify the tag is back in the suggested tags
    expect(screen.getByText('tag1')).toBeInTheDocument();
    expect(screen.getByText('Suggested Tags:')).toBeInTheDocument();
  });

  it('should call onFilterChats when tags are selected or removed', async () => {
    const mockChats: OptionType[] = [
      { value: '1', label: 'Chat 1', tags: ['tag1', 'tag2'] },
      { value: '2', label: 'Chat 2', tags: ['tag2', 'tag3'] }
    ];

    renderWithContext(<TagSearch {...defaultProps} />, { chats: mockChats });

    const input = screen.getByLabelText('Search for tags');
    fireEvent.change(input, { target: { value: 'tag' } });

    await waitFor(() => {
      fireEvent.click(screen.getByText('tag1'));
    });

    expect(mockOnFilterChats).toHaveBeenCalledWith([mockChats[0]]);

    fireEvent.click(screen.getByLabelText('Remove tag1 tag'));

    await waitFor(() => {
      expect(mockOnFilterChats).toHaveBeenCalledWith(mockChats);
    });
  });

  it('should clear search term after selecting a tag', async () => {
    const mockChats: OptionType[] = [
      { value: '1', label: 'Chat 1', tags: ['tag1', 'tag2'] }
    ];

    renderWithContext(<TagSearch {...defaultProps} />, { chats: mockChats });

    const input = screen.getByLabelText('Search for tags');
    fireEvent.change(input, { target: { value: 'tag' } });

    await waitFor(() => {
      fireEvent.click(screen.getByText('tag1'));
    });

    expect(input).toHaveValue('');
  });

  it('TagSearch component snapshot', () => {
    const { asFragment } = renderWithContext(<TagSearch {...defaultProps} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
