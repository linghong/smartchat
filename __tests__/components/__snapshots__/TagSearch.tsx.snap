// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TagSearch Component TagSearch component snapshot 1`] = `
<DocumentFragment>
  <div
    class="mb-6 bg-white p-4 rounded-lg shadow-md"
  >
    <label
      class="block text-sm font-medium text-gray-700 mb-2"
      for="tag-search"
    >
      Search Tags
    </label>
    <div
      class="relative"
    >
      <input
        aria-label="Search for tags"
        class="w-full px-4 py-2 border border-gray-300 rounded-md pl-10 text-sm focus:ring-2 focus:ring-blue-500 focus:border-blue-500"
        id="tag-search"
        placeholder="Type to search tags, and click a tag to select"
        type="text"
        value=""
      />
      <svg
        aria-hidden="true"
        class="lucide lucide-search absolute left-3 top-1/2 transform -translate-y-1/2 text-gray-400"
        fill="none"
        height="16"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="16"
        xmlns="http://www.w3.org/2000/svg"
      >
        <circle
          cx="11"
          cy="11"
          r="8"
        />
        <path
          d="m21 21-4.3-4.3"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
